rgw/CloudTransition: Transition objects to cloud endpoint

soumyakoduri opened the pull request:
This PR contains series of patches for CloudTiering/CloudTransition feature support (Experimental)

Feature Doc:
https://docs.google.com/document/d/1IoeITPCF64A5W-UA-9Y3Vp2oSfz3xVQHu31GTu3u3Ug/edit#heading=h.4o1xyas8s23x

Tier-config Options:
* Tier storage class and the respective remote Cloud endpoint details can be configured via "zonegroup placement add <>" CLI. 
* Existing tier-type & tier-config options are used to configure the same. Below are the config options made available -

    {
      "access_key": <access>,
      "secret": <secret>,
      "endpoint": <endpoint>,
     "region": "",
      "host_style": <path | virtual>,
      "acls": [ { "type": <id | email | uri>,
                  "source_id": <source_id>,
                  "dest_id": <dest_id> } ... ],
      "target_path": <target_path>,
      "target_storage_class": <target-storage-class>,
      "multipart_sync_threshold": {object_size},
      "multipart_min_part_size": {part_size},
      "retain_object": <true | false>
    }

Object Transition:
1) Check whether the target_placement is tier-storage class
2) If yes, depending on object size, transition object as plain object or in multiparts.
- Few custom rgw xattrs will be added to target object (eg., x-amz-meta-rgwx-source: rgw)
3) Post transition, either delete the original source object (default behavior) or mark the object as CloudTiered and update its head object Manifest with cloud-tier details.

Multisite:
To avoid multiple zones transitioning same object, each lc thread checks whether the target endpoint (cloud) already contains the object to be transitioned by issuing HEAD request and matching its mtime attr.

Test cases:
https://github.com/ceph/s3-tests/pull/380

TODO:
* Consolidate config and other common coroutine code with cloud-sync module.

To be handled in future:
* Send presigned url or read-through for GetObject  on cloud transitioned object.
* Restore operation/policies to restore tiered object
* federation between RGW and cloud/AWS

Signed-off-by: Soumya Koduri <skoduri@redhat.com>

soumyakoduri commented:
@yehudasa @mattbenjamin @cbodley @dang .. I have listed possible ways of configuring these cloud tier options (mainly to choose the right granularity - zonegroup/zone level) here - https://docs.google.com/document/d/1IoeITPCF64A5W-UA-9Y3Vp2oSfz3xVQHu31GTu3u3Ug/edit#heading=h.66scb8vgvd90 

Please take a look and let me know your suggestions on which one seems fit for this use-case and if any changes/additions are required. Thanks!

soumyakoduri commented:
unstale

soumyakoduri commented:
jenkins render docs

soumyakoduri commented:
jenkins render docs

ceph-jenkins commented:
Doc render available at https://ceph--35100.org.readthedocs.build/en/35100/

soumyakoduri commented:
jenkins render docs

ceph-jenkins commented:
Doc render available at https://ceph--35100.org.readthedocs.build/en/35100/

soumyakoduri commented:
jenkins render docs

ceph-jenkins commented:
Doc render available at https://ceph--35100.org.readthedocs.build/en/35100/

yehudasa reviewed (commented):
not done reviewing yet

soumyakoduri commented on doc/radosgw/cloud-transition.rst:
```diff
@@ -0,0 +1,325 @@
+                                               --placement-id={placement-id} \
+                                               --storage-class={storage-class-name} \
+                                               --tier-config=acls[].source_id=${source-id}, \
+                                               acls[${source-id}].dest_id=${dest-id}, \
```
${source-id} is the configured as source-id. For eg:

# bin/radosgw-admin zonegroup placement add --rgw-zonegroup default --placement-id default-placement --storage-class CLOUDTIER --tier-type=cloud --tier-config=acls[].source_id=testid12,acls[testid12].dest_id=rgw12,acls[testid12].type=email

For some reason, acls[-1] never worked  for me-- 

# bin/radosgw-admin zonegroup placement add --rgw-zonegroup default --placement-id default-placement --storage-class CLOUDTIER --tier-type=cloud --tier-config=acls[].source_id=testid13,acls[-1].dest_id=rgw123,acls[-1].type=email
2021-01-22T23:26:21.982+0530 7f0726b88940 -1 WARNING: all dangerous and experimental features are enabled.
2021-01-22T23:26:21.986+0530 7f0726b88940 -1 WARNING: all dangerous and experimental features are enabled.
2021-01-22T23:26:21.986+0530 7f0726b88940 -1 WARNING: all dangerous and experimental features are enabled.
ERROR: failed to set configurable: acls[-1].dest_id,rgw123
[root@localhost build]#

yehudasa replied:
Maybe need to do it in two separate commands? first add (` --tier-config=acls[].source_id=testid13`), then modify (`--tier-config=acls[-1].dest_id=rgw123,acls[-1].type=email`)

soumyakoduri commented on src/rgw/rgw_admin.cc:
```diff
@@ -4837,6 +4837,57 @@ int main(int argc, const char **argv)
+          RGWZoneGroupPlacementTier *pt = &t;
+
+	  auto ptiter = target.tier_targets.find(storage_class);
+	  if (ptiter != target.tier_targets.end()) {
```
done

soumyakoduri commented on doc/radosgw/cloud-transition.rst:
```diff
@@ -0,0 +1,325 @@
+
+See `Adding a Storage Class <https://docs.ceph.com/en/latest/radosgw/placement/#adding-a-storage-class>`__ for how to configure storage-class for a zonegroup. The cloud transition requires a creation of a special storage class with tier type defined as ``cloud``
+
+Note: Once a storage class is created of ``-tier-type=cloud``, it cannot be later modified to any other storage class type.
```
Done

soumyakoduri commented on src/rgw/rgw_zone.h:
```diff
@@ -675,10 +675,131 @@ struct RGWDefaultZoneGroupInfo {
+  void encode(bufferlist& bl) const {
+    ENCODE_START(1, 1, bl);
+    string s;
+    switch (type) {
```
done

soumyakoduri commented on src/rgw/rgw_zone.h:
```diff
@@ -675,10 +675,131 @@ struct RGWDefaultZoneGroupInfo {
+    DECODE_START(1, bl);
+      string s;
+     decode(s, bl);
+     if (s == "email") {
```
done

soumyakoduri commented on src/rgw/rgw_admin.cc:
```diff
@@ -4837,6 +4837,57 @@ int main(int argc, const char **argv)
+             * Once set cant be reset. */
+            tier_class = true;
+            pt->tier_type = tier_type;
+          }
```
done

yehudasa commented on src/rgw/rgw_lc.cc:
```diff
@@ -1249,32 +1252,243 @@ class LCOpAction_Transition : public LCOpAction {
+   std::map<std::string, RGWZoneGroupPlacementTier>::const_iterator ttier;
+   ttier = target_rule.tier_targets.find(storage_class);
+   if (ttier != target_rule.tier_targets.end()) {
+       tier = ttier->second;
```
indentation off maybe? not sure

soumyakoduri replied:
done.

yehudasa commented on src/rgw/rgw_lc.cc:
```diff
@@ -1249,32 +1252,243 @@ class LCOpAction_Transition : public LCOpAction {
+ }
+
+  int delete_tier_obj(lc_op_ctx& oc, RGWLCCloudTierCtx& tier_ctx) {
+    int ret = -1;
```
by default maybe just set it to `0`?

soumyakoduri replied:
done.

yehudasa commented on src/rgw/rgw_lc.cc:
```diff
@@ -1249,32 +1252,243 @@ class LCOpAction_Transition : public LCOpAction {
+    ret = crs.run(new RGWLCCloudCheckCR(tier_ctx, &al_tiered));
+    
+    if (ret < 0) {
+      ldpp_dout(oc.dpp, 0) << "ERROR: failed in RGWCloudCheckCR() ret=" << ret << dendl;
```
not returning here on purpose?

soumyakoduri replied:
yes.. "RGWLCCloudCheckCR" verifies if the object is already transitioned to remote cloud (incase of multisite). Since its just best effort to avoid duplicate transition, I thought it may be fine to ignore any error here.

yehudasa replied:
Can you add a comment explaining that?

soumyakoduri replied:
sure..done.

yehudasa commented on src/rgw/rgw_lc.cc:
```diff
@@ -1249,32 +1252,243 @@ class LCOpAction_Transition : public LCOpAction {
-      return r;
+    if (!r && oc.tier.tier_type == "cloud") {
+         ldpp_dout(oc.dpp, 0) << "Found cloud tier: " << target_placement.storage_class << dendl;
+        r = transition_obj_to_cloud(oc);
```
indentation

soumyakoduri replied:
done.

yehudasa commented on src/rgw/rgw_lc_tier.cc:
```diff
@@ -0,0 +1,1186 @@
+        }
+
+        grantee = iter->second.dest_id;
+
```
I see some partial similarities with the code at cloud sync, what prevented you from reusing same code? Maybe there's an opening for consolidation?

soumyakoduri replied:
@yehudasa  ..yes..along with config, most of the coroutine processing code is also similar. At the beginning of this project, I had bought this up in one of the refactoring call that if we can move most of the cloud sync code to a common place..but you had suggested that it shall become a lot more work than this feature actually is and you may be already working on refactoring that code. Sorry if I have misunderstood that part. 
If you suggest, I can work on consolidating both the modules code, atleast config part now.

mattbenjamin replied:
Let's try to strike a balance (easy to see path to consolidation, result not likely to be brittle, not too much longer to completion of this phase...)?

soumyakoduri replied:
have marked it as TODO - both code consolidation of the cloud-sync & cloud-transition modules and updating them with Zipper APIs. Will work on it once these changes get merged as it helps with testing too.

yehudasa commented on src/rgw/rgw_obj_manifest.h:
```diff
@@ -419,6 +459,38 @@ class RGWObjManifest {
     return max_head_size;
   }
 
+  string get_tier_type() {
```
`const` method

yehudasa commented on src/rgw/rgw_obj_manifest.h:
```diff
@@ -419,6 +459,38 @@ class RGWObjManifest {
+      tier_config.is_multipart_upload = t.is_multipart_upload;
+  }
+
+  void get_tier_config(RGWObjTier* t) {
```
`const` method

soumyakoduri replied:
done.

yehudasa commented on src/rgw/rgw_op.cc:
```diff
@@ -2129,6 +2129,18 @@ void RGWGetObj::execute(optional_yield y)
+    RGWObjManifest m;
+    decode(m, attr_iter->second);
+    if (m.get_tier_type() == "cloud") {
+      op_ret = -ERR_INVALID_OBJECT_STATE;
```
here you could in theory add the redirect code (if configured for such)

soumyakoduri replied:
thanks! I have added it as TODO comment for now.

yehudasa reviewed (commented):
see more comments

mattbenjamin commented on doc/radosgw/cloud-transition.rst:
```diff
@@ -0,0 +1,325 @@
+   x-amz-meta-rgwx-source-mtime: 1608546349.757100363
+   x-amz-meta-rgwx-versioned-epoch: 0
+
+By default, post transition, the source object gets deleted. But it is possible to retain its metadata but with updated values (like storage-class and object-size) by setting config option 'retain_object' to true. However GET on those objects shall still fail with 'InvalidObjectState' error.
```
I think we agreed that redirect or read-through would most likely be follow-on projects?

soumyakoduri replied:
yes.. This comment was earlier to that.. I shall add the same in the doc.

soumyakoduri replied:
added it to future work in the doc.

mattbenjamin replied:
Two questions.  First, this still seems ambiguous to me?  Delete implies ordinary deletion, which would be deletion from the source bucket index as well as on disk.  Is that currently the case?  Second, Casey mentioned you asked about object lock during transition.  I would vote that except in special circumstances (and by this I mean, to be defined later, I'm not sure they're well defined yet), we prohibit transition when an object is under object lock (retention or otherwise).  What do you think?

soumyakoduri replied:
delete (default case) here refers to deletion from both bucket index and disk.

yes. Object lock/retention is not currently handled. I too vote that it may be better to prohibit transition to cloud (as special case) if object is locked. 
I have one doubt. In the code currently, object_lock_retention checks ('pass_object_lock_check()') are done only for "LCOpAction_NonCurrentExpiration" case, but not for "LCOpAction_CurrentExpiration". Is this not applicable to current objects?

soumyakoduri replied:
Done. As discussed over irc, if its current version object which is transitioned to cloud, a delete marker is created for it . And non-current version locked objects are skipped from transition. Updated the doc too. Kindly review.

soumyakoduri commented on src/rgw/rgw_json_enc.cc:
```diff
@@ -97,6 +97,13 @@ void RGWObjManifestRule::dump(Formatter *f) const
 
+void RGWObjTier::dump(Formatter *f) const
+{
+  f->dump_string("name", name);
```
done.

soumyakoduri commented on src/rgw/rgw_lc.cc:
```diff
@@ -1249,32 +1252,243 @@ class LCOpAction_Transition : public LCOpAction {
+   std::map<std::string, RGWZoneGroupPlacementTarget>::const_iterator titer;
+   titer = zonegroup.placement_targets.find(rule.name);
+   if (titer == zonegroup.placement_targets.end()) {
+     return -1;
```
done.

soumyakoduri commented on src/rgw/rgw_json_enc.cc:
```diff
@@ -1394,11 +1406,88 @@ void RGWZone::decode_json(JSONObj *obj)
   JSONDecoder::decode_json("redirect_zone", redirect_zone, obj);
 }
 
+void RGWTierACLMapping::dump(Formatter *f) const
```
I am planning to address this in separate PR (once these changes are merged) as it helps with testing both the modules and fixing any other gaps too.

soumyakoduri commented on src/rgw/rgw_json_enc.cc:
```diff
@@ -1394,11 +1406,88 @@ void RGWZone::decode_json(JSONObj *obj)
+  JSONDecoder::decode_json("dest_id", dest_id, obj);
+}
+
+void RGWZoneGroupPlacementTier::dump(Formatter *f) const
```
have marked it as TODO - code consolidation of the cloud-sync & cloud-transition modules along with updating them with Zipper APIs .

soumyakoduri commented on doc/radosgw/cloud-transition.rst:
```diff
@@ -0,0 +1,331 @@
+                    "CLOUDTIER",
+                    "STANDARD"
+                ],
+                "tier_targets": [
```
@cbodley @yehudasa  ... As per your comment in the last refactoring call, changed the supported tier-type to 'cloud-s3'. Below are the options available for this tier-type, which can be configured as <key,value> pair (line: 175).  Kindly review.

mattbenjamin commented:
@soumyakoduri added comment on object lock + transition

mattbenjamin commented on src/rgw/rgw_lc.cc:
```diff
@@ -1249,32 +1252,244 @@ class LCOpAction_Transition : public LCOpAction {
+      ldpp_dout(oc.dpp, 0) << "failed in http_manager.start() ret=" << ret << dendl;
+      return ret;
+    }
+
```
I'm pretty sure there is;  can't we keep a timed cache of them around somewhere, linked from RGWLC?

cbodley replied:
agreed that we'll want to address this, especially because `RGWHTTPManager::start()` is spawning a background thread each time. it would be great if we could just use the global `RGWHTTPManager`, but we need one that's coupled to the `RGWCompletionManager` of this `RGWCoroutinesManager`

soumyakoduri replied:
Okay.. To address this, created a global http_manager  & crs (specific to
RGWLC). This http_manager & crs will be initialized and started  (if not already done) the first time an entry of a bucket is
being transitioned to the cloud . To avoid keeping them active forever, they are released post processing all the lc entries in that bucket.

Please let me know if this sounds fine.

soumyakoduri commented:
jenkins render docs

soumyakoduri commented:
jenkins render docs

ceph-jenkins commented:
Doc render available at https://ceph--35100.org.readthedocs.build/en/35100/

soumyakoduri commented:
teuthology run :https://pulpito.ceph.com/soumyakoduri-2021-03-23_07:25:50-rgw-cloudtiering-distro-basic-smithi/

soumyakoduri commented on src/rgw/rgw_cr_rest.h:
```diff
@@ -9,6 +9,7 @@
 
 #include "rgw_coroutine.h"
 #include "rgw_rest_conn.h"
+#include "rgw_rados.h"
```
It was needed to use ReadOp. Now with Zipper changes, have dependency only on "rgw_sal.h". Please let me know if thats fine.

soumyakoduri commented on src/rgw/rgw_lc.cc:
```diff
@@ -1249,32 +1251,273 @@ class LCOpAction_Transition : public LCOpAction {
+        ldpp_dout(oc.dpp, 20) << "delete_tier_obj Object(key:" << oc.o.key << ") current & not delete_marker" << "s versioned_epoch:  " << oc.o.versioned_epoch << "flags: " << oc.o.flags << dendl;
+    } else {
+        ret = remove_expired_obj(oc.dpp, oc, true);
+        ldpp_dout(oc.dpp, 20) << "delete_tier_obj Object(key:" << oc.o.key << ") not current " << "s versioned_epoch:  " << oc.o.versioned_epoch << "flags: " << oc.o.flags << dendl;
```
done.

soumyakoduri commented on src/rgw/rgw_lc.cc:
```diff
@@ -1249,32 +1251,273 @@ class LCOpAction_Transition : public LCOpAction {
+
+    r = obj_op->prepare(null_yield);
+
+    r = obj_op->write_meta(oc.dpp, tier_ctx.o.meta.size, 0, null_yield);
```
done.

soumyakoduri commented on do_cmake.sh:
```diff
@@ -14,7 +14,7 @@ if [ -e $BUILD_DIR ]; then
 
 PYBUILD="3"
-ARGS="-GNinja"
+#ARGS="-GNinja"
```
sorry..got added by mistake.

soumyakoduri commented on src/rgw/rgw_zone.h:
```diff
@@ -675,10 +675,153 @@ struct RGWDefaultZoneGroupInfo {
+    ENCODE_START(1, 1, bl);
+
+    string s = get_acl_type_str(type);
+    encode(s, bl);
```
done..updated to int.

soumyakoduri commented on src/rgw/rgw_zone.h:
```diff
@@ -675,10 +675,153 @@ struct RGWDefaultZoneGroupInfo {
+    encode(key, bl);
+    encode(region, bl);
+    string s = (host_style == PathStyle ? "path" : "virtual");
+    encode(s, bl);
```
done.

soumyakoduri commented on src/rgw/rgw_obj_manifest.h:
```diff
@@ -419,6 +459,38 @@ class RGWObjManifest {
     return max_head_size;
   }
 
+  const string get_tier_type() {
```
done.

soumyakoduri commented on src/rgw/rgw_obj_manifest.cc:
```diff
@@ -357,6 +357,8 @@ int RGWObjManifest::generator::create_begin(CephContext *cct, RGWObjManifest *_m
   manifest->set_tail_instance(_obj.key.instance);
 
+  manifest->set_tier_type("none");
+
```
right. Have updated it to be empty (default )

cbodley commented on doc/radosgw/cloud-transition.rst:
```diff
@@ -0,0 +1,347 @@
+
+This feature enables data transition to a remote cloud service as part of `Lifecycle Configuration <https://docs.aws.amazon.com/AmazonS3/latest/dev/object-lifecycle-mgmt.html>`__ via `Storage Classes <https://docs.ceph.com/en/latest/radosgw/placement/#storage-classes>`__. The transition is unidirectional; data cannot be transitioned back from the remote zone. The goal of this feature is to enable ...
+
+Special storage class of tier type ``cloud-s3`` is used to configure the remote cloud S3 object store service to which the data needs to be transitioned to. These are defined in terms of zonegroup placement targets and unlike regular storage classes, do not need a data pool. Any additions or modifications need period commit to get reflected.
```
> the remote cloud S3 object store service to which the data needs to be transitioned to.

extra 'to' at the end of the sentence

cbodley replied:
> Any additions or modifications need period commit to get reflected.

it's a little more complicated than that, unfortunately. the details are spelled out on the placement page:
> Note: If you have not done any previous Multisite Configuration, a default zone and zonegroup are created for you, and changes to the zone/zonegroup will not take effect until the Ceph Object Gateways are restarted. If you have created a realm for multisite, the zone/zonegroup changes will take effect once the changes are committed with radosgw-admin period update --commit.

you might want to copy that note somewhere under the 'How to Configure' section?

soumyakoduri replied:
done.

cbodley commented on doc/radosgw/cloud-transition.rst:
```diff
@@ -0,0 +1,347 @@
+How to Configure
+~~~~~~~~~~~~~~~~
+
+See `Adding a Storage Class <https://docs.ceph.com/en/latest/radosgw/placement/#adding-a-storage-class>`__ for how to configure storage-class for a zonegroup. The cloud transition requires a creation of a special storage class with tier type defined as ``cloud-s3``
```
another link that should be relative. you probably need to add a new label for it in placement.rst

cbodley commented on doc/radosgw/cloud-transition.rst:
```diff
@@ -0,0 +1,347 @@
+Cloud Transition
+================
+
+This feature enables data transition to a remote cloud service as part of `Lifecycle Configuration <https://docs.aws.amazon.com/AmazonS3/latest/dev/object-lifecycle-mgmt.html>`__ via `Storage Classes <https://docs.ceph.com/en/latest/radosgw/placement/#storage-classes>`__. The transition is unidirectional; data cannot be transitioned back from the remote zone. The goal of this feature is to enable ...
```
we should use relative links instead of `https://docs.ceph.com/en/latest/radosgw`, because this will eventually become the 'pacific' or 'quincy' doc, and we don't want the link to take you to a newer release

there's already a `_storage_classes` label in placement.rst, and s3.rst links to it like this:
```
| **Storage Class**               | Supported       | See :ref:`storage_classes`             |
```
that renders as `See Storage Classes`. you can do the same here

cbodley commented on src/rgw/rgw_acl.h:
```diff
@@ -46,6 +46,9 @@ enum ACLGroupTypeEnum {
 
+string get_acl_type_str (const ACLGranteeTypeEnum& type);
+ACLGranteeTypeEnum get_acl_type (const string& t);
+
```
it looks like these are unused now that the encoding changed?

soumyakoduri replied:
Done. get_acl_type() is still used but as there is only one caller, I have moved it inline of that routine now.

cbodley commented on src/rgw/rgw_lc_tier.cc:
```diff
@@ -0,0 +1,1242 @@
+static void init_headers(map<string, bufferlist>& attrs,
+                         map<string, string>& headers)
+{
+  for (auto kv : attrs) {
```
`const auto& kv` or we'll make a copy of each string and bufferlist

soumyakoduri replied:
done

soumyakoduri commented:
teuthology run : https://pulpito.ceph.com/soumyakoduri-2021-03-31_13:20:57-rgw-cloudtiering-distro-basic-smithi/

soumyakoduri commented:
jenkins render docs

ceph-jenkins commented:
Doc render available at https://ceph--35100.org.readthedocs.build/en/35100/

soumyakoduri commented:
retest this please

cbodley commented on src/rgw/rgw_lc_tier.cc:
```diff
@@ -0,0 +1,1242 @@
+  return 0;
+}
+
+map <pair<string, string>, utime_t> target_buckets;
```
a global variable? does anything ever expire these entries over the lifetime of the process?

soumyakoduri replied:
yes this is to avoid sending bucket creation request for every object processed (unlike in cloud-sync case). The entries are considered expired post '2*lc_debug_interval' time interval. Hope that will suffice. 
However there is no background thread to delete such entries.

cbodley commented on src/rgw/rgw_lc_tier.cc:
```diff
@@ -0,0 +1,1242 @@
+int RGWLCCloudCheckCR::operate() {
+  /* Check if object has already been transitioned */
+  rgw_lc_obj_properties obj_properties(tier_ctx.o.meta.mtime, tier_ctx.o.meta.etag,
+                                       tier_ctx.o.versioned_epoch, tier_ctx.acl_mappings,
+                                       tier_ctx.target_storage_class);
+
+  RGWBucketInfo b;
+  string target_obj_name;
+  int reterr = 0;
+
+  b.bucket.name = tier_ctx.target_bucket_name;
+  target_obj_name = tier_ctx.bucket_info.bucket.name + "/" +
+                    (*tier_ctx.obj)->get_name();
+    if (!tier_ctx.o.is_current()) {
+      target_obj_name += get_key_instance((*tier_ctx.obj)->get_key());
+    }
+
+  std::unique_ptr<rgw::sal::RGWBucket> dest_bucket;
+  std::unique_ptr<rgw::sal::RGWObject> dest_obj;
+
+
+  reterr = tier_ctx.store->get_bucket(nullptr, b, &dest_bucket);
+  if (reterr < 0) {
+    ldout(tier_ctx.cct, 0) << "ERROR: failed to initialize dest_bucket - " << tier_ctx.target_bucket_name << " , reterr = " << reterr << dendl;
+    return reterr;
+  }
+  
+  dest_obj = dest_bucket->get_object(rgw_obj_key(target_obj_name));
+  if (!dest_obj) {
+    ldout(tier_ctx.cct, 0) << "ERROR: failed to initialize dest_object path - " << target_obj_name << dendl;
+    return -1;
+  }
+
+  std::shared_ptr<RGWLCStreamGetCRF> get_crf;
+  get_crf.reset(new RGWLCStreamGetCRF((CephContext *)(tier_ctx.cct), get_env(), this,
+        (RGWHTTPManager*)(tier_ctx.http_manager),
+        obj_properties, tier_ctx.conn, static_cast<rgw::sal::RGWObject *>(dest_obj.get())));
+  int ret = 0;
+
+  reenter(this) {
```
note that every time this `RGWLCCloudCheckCR` coroutine yields, it will resume with a new call to `operate()`. we use the `reenter` macro, so that we can resume from where we left off

by placing all of this code outside of the `reenter` block, that means it gets re-executed every time we resume. it looks like all of this should move inside, so it only runs once at the beginning?

reenter is implemented with a switch statement, which means that local variables can be an issue; generally you need to make them member variables of the coroutine class instead

soumyakoduri replied:
done..

soumyakoduri replied:
I applied these changes to other coroutines as well and that fixed couple of issues I was facing with yield. Thanks! Updated the PR with latest code. Please review.

cbodley commented on src/rgw/rgw_lc_tier.cc:
```diff
@@ -0,0 +1,1242 @@
+  int ret = 0;
+
+  reenter(this) {
+    /* Having yield here doesn't seem to wait for init2() to fetch the headers
+     * before calling is_already_tiered() below
+     */
+    ret = get_crf->init();
```
can we figure out why this isn't working? maybe because all of the variables it relied on went out of scope because of the above comment?

without any yields, this code is completely synchronous, so all of the complexity from RGWCoroutine is wasted

soumyakoduri replied:
thats true. With the above changes, this issue got resolved. Thanks for the details.

cbodley commented on src/rgw/rgw_lc_tier.cc:
```diff
@@ -0,0 +1,1242 @@
+  }
+
+  std::shared_ptr<RGWLCStreamGetCRF> get_crf;
+  get_crf.reset(new RGWLCStreamGetCRF((CephContext *)(tier_ctx.cct), get_env(), this,
+        (RGWHTTPManager*)(tier_ctx.http_manager),
+        obj_properties, tier_ctx.conn, static_cast<rgw::sal::RGWObject *>(dest_obj.get())));
```
could this `shared_ptr` be `unique_ptr`, or even better just a member variable?

all of these casts to CephContext/RGWHTTPManager/rgw::sal::RGWObject look unnecessary

soumyakoduri replied:
done.

cbodley commented on src/rgw/rgw_lc_tier.cc:
```diff
@@ -0,0 +1,1242 @@
+      ldout(tier_ctx.cct, 0) << "ERROR: failed to fetch HEAD from cloud for obj=" << tier_ctx.obj << " , ret = " << ret << dendl;
+      return set_cr_error(ret);
+    }
+    if ((static_cast<RGWLCStreamGetCRF *>(get_crf.get()))->is_already_tiered()) {
```
why the cast, when the type is already `std::shared_ptr<RGWLCStreamGetCRF>`?

soumyakoduri replied:
addressed it.

soumyakoduri replied:
I applied these changes to other coroutines as well and that fixed couple of issues I was facing with yield. Thanks! Updated the PR with latest code. Please review.

soumyakoduri commented:
teuthology run - http://pulpito.front.sepia.ceph.com:80/soumyakoduri-2021-04-08_11:48:46-rgw-cloudtiering-distro-basic-smithi/

cbodley commented on src/rgw/rgw_lc_tier.cc:
```diff
@@ -568,42 +568,42 @@ class RGWLCStreamObjToCloudMultipartPartCR : public RGWCoroutine {
+  rgw_lc_obj_properties obj_properties;
+  RGWBucketInfo b;
+  string target_obj_name;
+  int ret = 0;
```
there's already a `RGWCoroutine::retcode` that you can use instead of `ret`. the only thing special about `retcode` is that it gets set automatically with the result of a `yield call`

soumyakoduri replied:
done.

cbodley commented on src/rgw/rgw_lc_tier.cc:
```diff
@@ -920,50 +920,48 @@ class RGWLCStreamObjToCloudMultipartCR : public RGWCoroutine {
+          "lc_multipart_" + (*tier_ctx.obj)->get_oid());
 
-      ret_err = tier_ctx.store->get_system_obj(tier_ctx.dpp, status_obj.pool,
+      ret = tier_ctx.store->get_system_obj(tier_ctx.dpp, status_obj.pool,
```
this coroutine is calling `get_system_obj()` `put_system_obj()` and `delete_system_obj()`, all of which are synchronous. you'll want to call the async versions instead: `RGWSimpleRadosReadCR`, `RGWSimpleRadosWriteCR`, and `RGWRadosRemoveCR`

soumyakoduri replied:
Initially had used RGWSimpleRados*CR but there is no abstraction for RGWAsyncRadosProcessor made part of Zipper 
 yet and hence couldn't use it directly. 

@dang mentioned that this may be addressed when sysobj is replaced in Zipper.

soumyakoduri replied:
So until the APIs are available in common sal layer, do you suggest to explicitly typecast to RGWRadosStore and fetch async_rados() for now?

cbodley replied:
i don't think `RGWAsyncRadosProcessor` itself has any dependency on rados, it's just a thread pool that coroutines use to dispatch synchronous work, while the coroutines are themselves asynchronous (they yield until the thread pool passes back the result)

that said, using `RGWAsyncRadosProcessor` is a hack that just avoids having to actually make the logic asynchronous, and i'd much rather we kill that thread pool than reify it as a part of zipper

in the meantime, i think it's okay for this tiering to be rados-backend-only and require a `RGWRadosStore` instead of the generic `RGWStore`; we just have the pick the right place in lifecycle processing to try `dynamic_cast<RGWRadosStore>` and return `ENOTSUP` on null

soumyakoduri replied:
okay sure thanks. Will make the changes.

soumyakoduri commented on src/rgw/rgw_lc.h:
```diff
@@ -481,6 +481,9 @@ class RGWLC : public DoutPrefixProvider {
     std::mutex lock;
     std::condition_variable cond;
     WorkPool* workpool{nullptr};
+    std::shared_ptr<RGWCoroutinesManager> crs;
```
@mattbenjamin @cbodley  @yehudasa .. to avoid creation of http_manager for each object transition, I have included it to be part of worker class. It will be started by the worker thread the first time an object is being transitioned to cloud and stopped post processing all the entries in that bucket. Please let me know if it sounds fine.

mattbenjamin replied:
@soumyakoduri seems reasonable to me.

soumyakoduri commented:
http://pulpito.front.sepia.ceph.com/soumyakoduri-2021-04-27_17:08:20-rgw-cloudtiering-distro-basic-smithi/

soumyakoduri commented:
> > @soumyakoduri I've looked at the stop http_manager part. I suspect that destroying http_manager immediately whenever a workq is drained is too fast. I've tested workloads where lifecycle essentially ran continuously, and in these, workqs could enter and exit idle state very quickly. I think we would like a way to delete http_manager only if the workq remains idle for a while (e.g., start a delete timer on idle, but allow it to be cleared if the workq becomes active again before the timer expires?).
> 
> Okay..thanks..Will make the necessary changes.

@mattbenjamin  Updated the changes. Kindly review.

soumyakoduri commented:
https://pulpito.ceph.com/soumyakoduri-2021-05-13_16:46:38-rgw-cloudtiering-distro-basic-smithi/

mattbenjamin commented on src/rgw/rgw_lc.cc:
```diff
@@ -777,6 +786,10 @@ class WorkQ : public Thread
       if (flags & FLAG_EDRAIN_SYNC) {
 	flags &= ~FLAG_EDRAIN_SYNC;
       }
+      if (flags & FLAG_HTTP_MGR) {
```
ok

mattbenjamin commented on src/rgw/rgw_lc.cc:
```diff
@@ -746,6 +713,48 @@ class WorkQ : public Thread
+      return 0;
+    }
+
+    http_manager.reset();
```
I didn't know you could call reset with no argument

soumyakoduri commented on src/rgw/rgw_lc.cc:
```diff
@@ -1271,22 +1273,68 @@ class LCOpAction_Transition : public LCOpAction {
    return 0;
  }
+  int transition_obj_to_cloud(lc_op_ctx& oc) {
+    std::shared_ptr<RGWRESTConn> conn;
```
Its initialized with S3RESTConn() below at line:1290

soumyakoduri commented on src/rgw/rgw_lc_tier.cc:
```diff
@@ -0,0 +1,1035 @@
+    /* New attribute to specify its transitioned from RGW */
+    new_attrs["x-amz-meta-rgwx-source"] = "rgw";
+
+    char buf[32];
```
ok thanks

soumyakoduri commented on src/rgw/rgw_lc.cc:
```diff
@@ -1336,7 +1347,9 @@ class LCOpAction_Transition : public LCOpAction {
+     * obj_size remains the same even when object is moved to other
+     * storage class. So maybe better to keep it the same way.
      */
     //pmanifest->set_obj_size(0);
```
obj_size is preserved here. Will remove that comment.

soumyakoduri commented on src/rgw/rgw_acl.cc:
```diff
@@ -74,6 +74,36 @@ bool operator!=(const RGWAccessControlPolicy& lhs,
+  return s;
+}
+
+ACLGranteeTypeEnum get_acl_type (const string& t) {
```
done.

soumyakoduri commented on src/rgw/rgw_acl.cc:
```diff
@@ -74,6 +74,36 @@ bool operator!=(const RGWAccessControlPolicy& lhs,
   return !(lhs == rhs);
 }
 
+string get_acl_type_str (const ACLGranteeTypeEnum& type) {
```
done.

soumyakoduri commented on src/rgw/rgw_cr_rest.cc:
```diff
@@ -347,3 +347,105 @@ int RGWStreamSpliceCR::operate(const DoutPrefixProvider *dpp) {
+RGWStreamWriteCR::~RGWStreamWriteCR() { }
+
+int RGWStreamWriteCR::operate() {
+  off_t ofs;
```
This is outdated code.

soumyakoduri commented on src/rgw/rgw_cr_rest.cc:
```diff
@@ -347,3 +347,105 @@ int RGWStreamSpliceCR::operate(const DoutPrefixProvider *dpp) {
+      op_target(rados, bucket_info, obj_ctx, obj), read_op(&op_target) {}
+RGWStreamReadCRF::~RGWStreamReadCRF() {}
+
+RGWStreamWriteCR::RGWStreamWriteCR(CephContext *_cct, RGWHTTPManager *_mgr,
```
done.

soumyakoduri commented on src/rgw/rgw_lc_tier.cc:
```diff
@@ -0,0 +1,1035 @@
+
+using namespace std;
+
+static string get_key_oid(const rgw_obj_key& key)
```
done.

soumyakoduri commented on src/rgw/rgw_lc_tier.cc:
```diff
@@ -0,0 +1,1035 @@
+    rgw_rest_obj rest_obj;
+
+    reenter(this) {
+#ifdef TODO_STATUS_OBJ
```
This was outdated code.

soumyakoduri commented on src/rgw/rgw_lc_tier.cc:
```diff
@@ -96,6 +96,126 @@ static void init_headers(map<string, bufferlist>& attrs,
+    req_params.get_op = false; /* Need only headers */
+    req_params.prepend_metadata = true;
+
+//    req_params.unmod_ptr = &src_properties.mtime;
```
This was outdated code.

soumyakoduri commented on src/rgw/rgw_lc.cc:
```diff
@@ -1391,7 +1391,19 @@ class LCOpAction_Transition : public LCOpAction {
+    ret = crs.run(new RGWLCCloudCheckCR(tier_ctx, &al_tiered));
+    
+    if (ret < 0) {
+      ldpp_dout(oc.dpp, 0) << "XXXXXXXXXXXXXX failed in RGWCloudCheckCR() ret=" << ret << dendl;
```
This was outdated code.

soumyakoduri commented on src/rgw/rgw_lc.cc:
```diff
@@ -1445,14 +1458,14 @@ class LCOpAction_Transition : public LCOpAction {
 
-    RGWCoroutinesManager *crs = oc.env.worker->get_crs();
-    RGWHTTPManager *http_manager = oc.env.worker->get_http_manager();
+    RGWCoroutinesManager *crs = oc.wq->get_crs();
```
done.

soumyakoduri commented on src/rgw/rgw_lc.cc:
```diff
@@ -1249,32 +1251,262 @@ class LCOpAction_Transition : public LCOpAction {
+    target_placement.storage_class = oc.tier.storage_class;
+    pmanifest->set_head(target_placement, (*tier_ctx.obj)->get_obj(), 0);
+
+    pmanifest->set_tail_placement(target_placement, (*tier_ctx.obj)->get_obj().bucket);
```
@cbodley  please let me know if the updates to object manifest done here are sufficient? Is there a way to retain head obj but delete only tail objects from manifest->objs?

soumyakoduri commented:
jenkins test this please

soumyakoduri commented:
https://pulpito.ceph.com/soumyakoduri-2021-06-28_17:41:36-rgw-cloudtiering-distro-basic-smithi/

soumyakoduri commented:
https://pulpito.ceph.com/soumyakoduri-2021-10-17_08:45:34-rgw-cloudtiering-distro-basic-smithi/

cbodley commented on src/rgw/rgw_lc_tier.cc:
```diff
@@ -0,0 +1,1072 @@
+ { RGW_ATTR_CONTENT_ENC,       "Content-Encoding" },
+ { RGW_ATTR_USER_MANIFEST,     "X-Object-Manifest" },
+ { RGW_ATTR_X_ROBOTS_TAG ,     "X-Robots-Tag" },
+ { RGW_ATTR_STORAGE_CLASS ,    "X-Amz-Storage-Class" },
```
don't we want to use the tier config's `target_storage_class` for this?

cbodley commented on src/rgw/rgw_lc_tier.cc:
```diff
@@ -0,0 +1,1072 @@
+      headers[aiter->second] = rgw_bl_str(kv.second);
+    } else if (strcmp(name, RGW_ATTR_SLO_UINDICATOR) == 0) {
+      // this attr has an extra length prefix from encode() in prior versions
+      headers["X-Object-Meta-Static-Large-Object"] = "True";
```
this is a swift thing, not s3

cbodley commented on src/rgw/rgw_lc_tier.cc:
```diff
@@ -0,0 +1,1072 @@
+          sizeof(RGW_ATTR_META_PREFIX)-1) == 0) {
+      name += sizeof(RGW_ATTR_META_PREFIX) - 1;
+      string sname(name);
+      string name_prefix = "X-Object-Meta-";
```
this looks like the swift format. user metadata in s3 starts with `x-amz-meta-`

cbodley commented on src/rgw/rgw_lc_tier.cc:
```diff
@@ -0,0 +1,1072 @@
+      headers[full_name_buf] = rgw_bl_str(kv.second);
+    } else if (strcmp(name,RGW_ATTR_CONTENT_TYPE) == 0) {
+      /* Verify if its right way to copy this field */
+      headers["CONTENT_TYPE"] = rgw_bl_str(kv.second);
```
> Verify if its right way to copy this field 

that looks fine

cbodley commented on src/rgw/rgw_lc_tier.cc:
```diff
@@ -0,0 +1,1072 @@
+      (long long)ut.sec(),
+      (long long)ut.nsec());
+
+  (*attrs)["x-amz-meta-rgwx-source-mtime"] = buf;
+  (*attrs)["x-amz-meta-rgwx-source-etag"] = obj_properties.etag;
+  (*attrs)["x-amz-meta-rgwx-source-key"] = rest_obj.key.name;
+  if (!rest_obj.key.instance.empty()) {
+    (*attrs)["x-amz-meta-rgwx-source-version-id"] = rest_obj.key.instance;
```
just pass `attrs` by reference instead of pointer

cbodley reviewed (commented):
i think the transfer logic under `RGWLCCloudTier` looks a lot better without the coroutines!

cbodley commented on src/rgw/rgw_lc_tier.cc:
```diff
@@ -0,0 +1,1072 @@
+  }
+
+  if (retcode >= 0) {
+    // check here that mtime and size did not change 
+    if (status.mtime != obj_properties.mtime || status.obj_size != obj_size ||
+        status.etag != obj_properties.etag) {
+      std::unique_ptr<RGWLCCloudStreamAbortMultipartUpload> abort_mp;
```
ok, so the sync status object helps us identify incomplete transfers that need to be aborted. but outside of `mtime`/`obj_size`/`etag`/`upload_id`, that sync status stores a lot of extra stuff that it probably doesn't need to

it doesn't appear to be necessary to rewrite the sync status with `put_sync_status()` every time we update `cur_part`, `cur_ofs`, or `parts`; all of those variables could probably just be on the stack instead of encoding them inside the sync status object with `rgw_lc_multipart_upload_info`

if this `RGWLCCloudStreamAbortMultipartUpload` case is the only reason we're persisting the sync status object, then we should only write it once (immediately after `RGWLCCloudInitMultipart` succeeds)

soumyakoduri replied:
@cbodley .... yes now that all these calls are synchronous and not part of coroutines, there doesn't seem any need to store sync status object . Could you please confirm if it can be removed completely.

cbodley replied:
i believe we still want it, so that we can abort previous attempts to upload:

> if this RGWLCCloudStreamAbortMultipartUpload case is the only reason we're persisting the sync status object, then we should only write it once (immediately after RGWLCCloudInitMultipart succeeds)

soumyakoduri replied:
okay thanks!

soumyakoduri commented on src/rgw/rgw_lc.cc:
```diff
@@ -1248,32 +1251,255 @@ class LCOpAction_Transition : public LCOpAction {
+    }
+
+    /* Create RGW REST connection */
+    conn.reset(new S3RESTConn(oc.cct, oc.store, id, { endpoint }, key, region, host_style));
```
done.

soumyakoduri commented on src/rgw/rgw_lc_tier.h:
```diff
@@ -0,0 +1,563 @@
+};
+
+/* Transition object to cloud endpoint */
+class RGWLCCloudTier {
+  RGWLCCloudTierCtx& tier_ctx;
+  bufferlist out_bl;
+  int retcode;
+  std::unique_ptr<RGWLCCloudCheckObj> chk_cloud;
+  std::unique_ptr<RGWLCCloudStreamObjToPlain> cloud_tier_plain;
+  std::unique_ptr<RGWLCCloudStreamObjToMultipart> cloud_tier_mp;
```
done.

soumyakoduri commented:
> i think the transfer logic under `RGWLCCloudTier` looks a lot better without the coroutines!

Thanks for the review. I addressed the comments and updated the PR. Please take a look.

cbodley commented on src/rgw/rgw_lc.cc:
```diff
@@ -1248,32 +1252,256 @@ class LCOpAction_Transition : public LCOpAction {
+    tier_ctx.storage_class = oc.tier.storage_class;
+
+    // check if target_path is already created
+    std::vector<std::string>& cloud_targets = oc.env.worker->get_cloud_targets();
+
+    std::vector<std::string>::iterator it = std::find(cloud_targets.begin(),
+                                                 cloud_targets.end(), bucket_name);
+    if (it != cloud_targets.end()) { //target path already created in this LC run
+      tier_ctx.create_target_bucket = false;
```
i think std::set is a better fit than std::vector here, because it gives you O(logN) search instead of O(N)

and instead of doing the search here, and separate insert after `rgw_cloud_tier_transfer_object()` returns, you can just call `set::insert()` directly and use the `bool inserted` flag it returns:
```c++
// try to create the target bucket if this is the first time we've seen bucket_name
tier_ctx.create_target_bucket = cloud_targets.insert(bucket_name).second;
```

cbodley reviewed (commented):
rgw_lc_tier.cc is looking much better!

cbodley commented on src/rgw/rgw_lc_tier.cc:
```diff
@@ -0,0 +1,1273 @@
+  (*tier_ctx.obj)->set_atomic(&tier_ctx.rctx);
+
+  /* Prepare Read from source */
+  std::shared_ptr<RGWLCStreamRead> readf;
+  readf.reset(new RGWLCStreamRead(tier_ctx.cct, tier_ctx.dpp,
+        tier_ctx.rctx, tier_ctx.obj, tier_ctx.o.meta.mtime));
+
+  std::shared_ptr<RGWLCCloudStreamPut> writef;
+  writef.reset(new RGWLCCloudStreamPut(tier_ctx.dpp, obj_properties, tier_ctx.conn,
+               dest_obj.get()));
+
+  /* actual Read & Write */
+  ret = cloud_tier_transfer_object(tier_ctx.dpp, readf.get(), writef.get());
```
ok, can we at least leave a `// TODO` comment about that?

soumyakoduri replied:
done.

soumyakoduri commented on src/rgw/rgw_lc_tier.h:
```diff
@@ -0,0 +1,147 @@
+#include "rgw_cr_rados.h"
+#include "rgw_rest_conn.h"
+#include "rgw_cr_rest.h"
+#include "rgw_coroutine.h"
```
'rgw_rest_obj' is defined in rgw_cr_rest.h. Please let me know if that needs to be moved to another common header.

soumyakoduri commented:
> what happens if we try to delete an object that's been cloud-tiered?

they will be deleted like any other regular object.

soumyakoduri commented:
jenkins test make check

cbodley commented:
> > what happens if we try to delete an object that's been cloud-tiered?
> 
> they will be deleted like any other regular object.

i'm curious whether that deletion tries to GC the original tail objects. i guess that behavior depends on what we left in the object's manifest

soumyakoduri commented:
> > > what happens if we try to delete an object that's been cloud-tiered?
> > 
> > 
> > they will be deleted like any other regular object.
> 
> i'm curious whether that deletion tries to GC the original tail objects. i guess that behavior depends on what we left in the object's manifest

yes..I assume original tail objects will be moved to  GC list as soon as the object is transitioned and we set obj_size to '0'.  I will verify that.

soumyakoduri commented:
> > > > what happens if we try to delete an object that's been cloud-tiered?
> > > 
> > > 
> > > they will be deleted like any other regular object.
> > 
> > 
> > i'm curious whether that deletion tries to GC the original tail objects. i guess that behavior depends on what we left in the object's manifest
> 
> yes..I assume original tail objects will be moved to GC list as soon as the object is transitioned and we set obj_size to '0'. I will verify that.

Creating a new manifest instead of reusing and updating the current one fixed this issue. It now deletes tail objects while updating the head object metadata post transition to cloud.

soumyakoduri commented:
jenkins test

soumyakoduri commented:
jenkins test make check
jenkins test api

soumyakoduri commented:
http://pulpito.front.sepia.ceph.com/soumyakoduri-2021-11-11_20:19:05-rgw-cloudtiering-distro-basic-smithi/

cbodley commented:
thanks @soumyakoduri, the results look good. there's a lot of commit history here, can you look through it and see if any of it can be easily squashed out? some of the 'fix rebase' and http manager commits are potential candidates. but if you find it's getting messy, don't spend much time on it :)

soumyakoduri commented:
> thanks @soumyakoduri, the results look good. there's a lot of commit history here, can you look through it and see if any of it can be easily squashed out? some of the 'fix rebase' and http manager commits are potential candidates. but if you find it's getting messy, don't spend much time on it :)

Done.

cbodley commented:
jenkins test dashboard cephadm

cbodley commented:
jenkins test dashboard cephadm

cbodley commented:
jenkins test windows
